### ISSUE TITLE
Commenting on Issues

### ISSUE THREAD

**Issue by marusinm** [packit-service/ogr#80]
Based on our discussion on Issue #79, there is no functionality in ogr for commenting on Github/Pagure issues.

**Comment by marusinm** [comment]
I would like to provide a solution to this issue. 

Before that, I have several questions about ogr development: 
1. when you creating such a feature like this, should one PR consists of functionality for all git forges at once. In this case, I would provide commenting on Issue for Github/Pagure/Gitlab in one PR.
2. What is current support for Gitlab? I looked at ogr/services/gitlab.py and for now, it seems to be unsupported.
3.  Is there any special way how to run tests? Or something I should know about writing them? *Release-bot*, for example, creates its own container for testing with a makefile.

**Comment by rpitonak** [comment]
> 1. when you creating such a feature like this, should one PR consists of functionality for all git forges at once. In this case, I would provide commenting on Issue for Github/Pagure/Gitlab in one PR.

I think this is up to you, small PRs are easier to review. About Gitlab, as far as I know, we focus mainly on Github and Pagure right now.

> 3. Is there any special way how to run tests? Or something I should know about writing them? _Release-bot_, for example, creates its own container for testing with a makefile.

@jscotka can provide the best answer to this question.

**Comment by TomasTomecek** [comment]
> I would like to provide a solution to this issue. 
> 
> Before that, I have several questions about ogr development: 
> 
>     1. when you creating such a feature like this, should one PR consists of functionality for all git forges at once. In this case, I would provide commenting on Issue for Github/Pagure/Gitlab in one PR.

Since none of us is really using the gitlab interface, I wouldn't worry with it much (unless you like adventure).

As Rado said, send the PR as you wish. One PR is okay.

>     2. What is current support for Gitlab? I looked at ogr/services/gitlab.py and for now, it seems to be unsupported.

Yup, no one is using it (yet).

>     3. Is there any special way how to run tests? Or something I should know about writing them? _Release-bot_, for example, creates its own container for testing with a makefile.

https://github.com/packit-service/ogr/blob/master/CONTRIBUTING.md#makefile

please let us know if the doc is outdated

**Comment by marusinm** [comment]
I successfully run tests on my local machine, now I wanted to test my new functionality for this Issue. 

However, I need to generate a similar file like .yaml files stored in tests/integration/test-data. Without that, I am not able to test new functionality. Any advice on how to generate such .yaml file?

**Comment by jscotka** [comment]
Hi, see lines in: https://github.com/packit-service/ogr/blob/master/tests/integration/test_github.py#L20
So for regenereation these files, you have to use syntax eg:
``FORCE_WRITE=yes GITHUB_TOKEN=yourtoken GITHUB_USER=yourname pytest3 tests/integration/test_github.py``
or this will also work ``FORCE_WRITE=yes GITHUB_TOKEN=yourtoken GITHUB_USER=yourname make check`` with make check, you have to have set PAGURE_TOKEN and USER env vars

**Comment by marusinm** [comment]
So I created `class Issues(GithubTests)` in `test_github.py`. This class should create one of the .yaml files similar to all files in tests/integration/test-data.

 In my case it should create for example file:
```
my_workspace/ogr/tests/integration/test_data/test_github_data_tests.integration.test_github.Issues.test_issue_list.yaml
```

However, when I run all tests, all of them passing except for my tests in `class Issues(GithubTests)`.  I got an exception:
```
ogr.exceptions.PersistenStorageException: file does not exists: my_workspace/ogr/tests/integration/test_data/test_github_data_tests.integration.test_github.Issues.test_issue_list.yaml
```

In `setUp` function within `class GithubTests` there are 3 env variables. 
1. `GITHUB_TOKEN` 
2. `GITHUB_USER`
3. `FORCE_WRITE`

First two are set inside my virtualenv. The third one is always `False` otherwise all tests in `test_github.py` start to fail.

Firstly I thought It is a problem with permissions because these tests are running on your repo `user-cont/colin`. However, when I tried my own repo with my permissions for my tests, they are still failing. 

Everything seems to be set right including environment variables, paths, permission. Can you help me to find what I'm doing wrong?
@jscotka

**Comment by marusinm** [comment]
So I created `class Issues(GithubTests)` in `test_github.py`. This class should create one of the .yaml files similar to all files in tests/integration/test-data.

 In my case it should create for example file:
```
my_workspace/ogr/tests/integration/test_data/test_github_data_tests.integration.test_github.Issues.test_issue_list.yaml
```

However, when I run all tests, all of them passing except for my tests in `class Issues(GithubTests)`.  I got an exception:
```
ogr.exceptions.PersistenStorageException: file does not exists: my_workspace/ogr/tests/integration/test_data/test_github_data_tests.integration.test_github.Issues.test_issue_list.yaml
```

In `setUp` function within `class GithubTests` there are 3 env variables. 
1. `GITHUB_TOKEN` 
2. `GITHUB_USER`
3. `FORCE_WRITE`

First two are set inside my virtualenv. The third one is always `False` otherwise all tests in `test_github.py` start to fail.

Firstly I thought It is a problem with permissions because these tests are running on your repo `user-cont/colin`. However, when I tried my own repo with my permissions for my tests, they are still failing. 

Everything seems to be set right including environment variables, paths, permission. Can you help me to find what I'm doing wrong?
@jscotka 

**Edit:**
Sorry I missed your last comment. It looks I did everything in the right way. However what is `FORCE_WRITE` env variable? When I set it to `True`/`yes`/`1` all tests are failing.

I'm also using macOS, can be this problem? I tried to run tests with `tox` and also `pytest` inside my virtualenv.

**Comment by lachmanfrantisek** [comment]
@marusinm 

When you set FORCE_WRITE to true, the tests will be directly requesting the GitHub/Pagure. The response will be saved in those YAML files.

When you do not use FORSE_WRITE, the tests will not use GitHub, but those YAML files to get a response... (This is used in the CI.)

Just create a pull-request and we can help you with that...

**Comment by jscotka** [comment]
exactly, with ``FORCE_WRITE`` (set to any nonempty string :-) enforce writing files and reading real github API and store requests. inside CI is not set so it just use files. tox file has also enable to pass this value https://github.com/packit-service/ogr/blob/master/tox.ini#L18 so that you can regenerate it also via tox, but in this case you have to regenerate everything so that pagure variables has to be set as well.

**Comment by lachmanfrantisek** [comment]
This is already solved by #99 and #81
